atmcfg/atmcfg.go

```diff
@@ -44,37 +56,118 @@ func setDisabledByReplicaSetName(out *opsmngr.AutomationConfig, name string, dis
 		for _, m := range rs.Members {
 			for k, p := range out.Processes {
 				if p.Name == m.Host {
-					out.Processes[k].Disabled = disabled
+					setDisable(out.Processes[k], processesMap, disabled)
 				}
 			}
 		}
 	}
 }
 
-func setDisabledByShardName(out *opsmngr.AutomationConfig, name string, disabled bool) {
+func setDisabledByShardNameAndProcesses(out *opsmngr.AutomationConfig, name string, processesMap map[string]bool, disabled bool) {
 	i, found := search.ShardingConfig(out.Sharding, func(s *opsmngr.ShardingConfig) bool {
 		return s.Name == name
 	})
 	if found {
 		s := out.Sharding[i]
 		// disable shards
 		for _, rs := range s.Shards {
-			setDisabledByReplicaSetName(out, rs.ID, disabled)
+			setDisabledByReplicaSetNameAndProcesses(out, rs.ID, processesMap, disabled)
 		}
 		// disable config rs
-		setDisabledByReplicaSetName(out, s.ConfigServerReplica, disabled)
+		setDisabledByReplicaSetNameAndProcesses(out, s.ConfigServerReplica, processesMap, disabled)
 		// disable mongos
 		for i := range out.Processes {
 			if out.Processes[i].Cluster == name {
-				out.Processes[i].Disabled = disabled
+				setDisable(out.Processes[i], processesMap, disabled)
 			}
 		}
 	}
 }
 
+func setDisable(process *opsmngr.Process, processesMap map[string]bool, disabled bool) {
+	if len(processesMap) == 0 {
+		process.Disabled = disabled
+		return
+	}
+
+	key := fmt.Sprintf("%s:%d", process.Hostname, process.Args26.NET.Port)
+	if _, ok := processesMap[key]; ok {
+		process.Disabled = disabled
+		processesMap[key] = true
+	}
+}
+
 // Shutdown disables all processes of the given cluster name.
-func Shutdown(out *opsmngr.AutomationConfig, name string) {
-	setDisabledByClusterName(out, name, true)
+func Shutdown(out *opsmngr.AutomationConfig, clusterName string) {
+	setDisabledByClusterName(out, clusterName, true)
+}
+
+// ShutdownProcessesByClusterName disables the entire cluster or its processes. Processes are provided in the format {"hostname:port","hostname2:port2"}.
+func ShutdownProcessesByClusterName(out *opsmngr.AutomationConfig, clusterName string, processes []string) error {
+	if len(processes) == 0 {
+		Shutdown(out, clusterName)
+		return nil
+	}
+	return shutdownProcesses(out, clusterName, processes)
+}
+
+// shutdownProcesses disables processes. Processes are provided in the format {"hostname:port","hostname2:port2"}.
+func shutdownProcesses(out *opsmngr.AutomationConfig, clusterName string, processes []string) error {
+	processesMap := map[string]bool{}
+	for _, hostnameAndPort := range processes {
+		processesMap[hostnameAndPort] = false
+	}
+
+	newDeploymentAuthMechanisms(out)
+	setDisabledByReplicaSetNameAndProcesses(out, clusterName, processesMap, true)
+	setDisabledByShardNameAndProcesses(out, clusterName, processesMap, true)
```

Comment:
[super nit]: I'd have made a helper to do this (replica set and sharded clusters) as there's a few places where you're checking both configs.

Not blocking/personal nit only)